From 4a6adf42d80a91f79c9c7c8d0c7801f9575d40dd Mon Sep 17 00:00:00 2001
From: Sharjeel Khan <sharjeelkhan@google.com>
Date: Tue, 7 Apr 2026 11:48:45 -0400
Subject: [PATCH] [InstSimplify] Fix Compilation Hang in
 simplifyExtractValueInst (#190279)

Jump Threading can create self-referential insertvalues which are
allowed by the verifier in unreachable code. These self-referential
insertvalues cause the compilation to hang in simplifyExtractValueInst.
This PR adds a check to break out of the loop if it detects it is a
self-referential insertvalue and adds the reproducer's bitcode as a
test.

Fixes: https://github.com/llvm/llvm-project/issues/187381
---
 llvm/lib/Analysis/InstructionSimplify.cpp     |  5 ++
 .../CorrelatedValuePropagation/pr187381.ll    | 58 +++++++++++++++++++
 2 files changed, 63 insertions(+)
 create mode 100644 llvm/test/Transforms/CorrelatedValuePropagation/pr187381.ll

diff --git a/llvm/lib/Analysis/InstructionSimplify.cpp b/llvm/lib/Analysis/InstructionSimplify.cpp
index 12c3edc0ced3..4b7e23d3025a 100644
--- a/llvm/lib/Analysis/InstructionSimplify.cpp
+++ b/llvm/lib/Analysis/InstructionSimplify.cpp
@@ -5455,8 +5455,13 @@ static Value *simplifyExtractValueInst(Value *Agg, ArrayRef<unsigned> Idxs,
 
   // extractvalue x, (insertvalue y, elt, n), n -> elt
   unsigned NumIdxs = Idxs.size();
+  SmallPtrSet<InsertValueInst *, 8> VisitedSet;
   for (auto *IVI = dyn_cast<InsertValueInst>(Agg); IVI != nullptr;
        IVI = dyn_cast<InsertValueInst>(IVI->getAggregateOperand())) {
+    // Protect against insertvalue cycles in unreachable code.
+    if (!VisitedSet.insert(IVI).second)
+      break;
+
     ArrayRef<unsigned> InsertValueIdxs = IVI->getIndices();
     unsigned NumInsertValueIdxs = InsertValueIdxs.size();
     unsigned NumCommonIdxs = std::min(NumInsertValueIdxs, NumIdxs);
diff --git a/llvm/test/Transforms/CorrelatedValuePropagation/pr187381.ll b/llvm/test/Transforms/CorrelatedValuePropagation/pr187381.ll
new file mode 100644
index 000000000000..80ddf0a6aedd
--- /dev/null
+++ b/llvm/test/Transforms/CorrelatedValuePropagation/pr187381.ll
@@ -0,0 +1,58 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 6
+; RUN: opt < %s -passes=correlated-propagation -S | FileCheck %s
+
+define i32 @insertvalue_selfreferential() {
+; CHECK-LABEL: define i32 @insertvalue_selfreferential() {
+; CHECK-NEXT:  [[ENTRY:.*]]:
+; CHECK-NEXT:    br label %[[EXIT:.*]]
+; CHECK:       [[UNREACHABLE_BLOCK:.*]]:
+; CHECK-NEXT:    [[TMP0:%.*]] = insertvalue [2 x i32] [[TMP0]], i32 0, 1
+; CHECK-NEXT:    [[EXTRACT_VALUE:%.*]] = extractvalue [2 x i32] [[TMP0]], 0
+; CHECK-NEXT:    br label %[[EXIT]]
+; CHECK:       [[EXIT]]:
+; CHECK-NEXT:    [[SINK_VALUE:%.*]] = phi i32 [ 0, %[[ENTRY]] ], [ [[EXTRACT_VALUE]], %[[UNREACHABLE_BLOCK]] ]
+; CHECK-NEXT:    ret i32 0
+;
+entry:
+  br label %exit
+
+unreachable.block:                                ; No predecessors!
+  %0 = insertvalue [2 x i32] %0, i32 0, 1
+  %extract.value = extractvalue [2 x i32] %0, 0
+  br label %exit
+
+exit:                             ; preds = %unreachable.block, %entry
+  %sink.value = phi i32 [ 0, %entry ], [ %extract.value, %unreachable.block ]
+  ret i32 0
+}
+
+define i32 @insertvalue_longchain() {
+; CHECK-LABEL: define i32 @insertvalue_longchain() {
+; CHECK-NEXT:  [[ENTRY:.*]]:
+; CHECK-NEXT:    br label %[[EXIT:.*]]
+; CHECK:       [[UNREACHABLE_BLOCK:.*]]:
+; CHECK-NEXT:    [[TMP0:%.*]] = insertvalue [2 x i32] [[TMP3:%.*]], i32 0, 1
+; CHECK-NEXT:    [[TMP1:%.*]] = insertvalue [2 x i32] [[TMP0]], i32 0, 1
+; CHECK-NEXT:    [[TMP2:%.*]] = insertvalue [2 x i32] [[TMP1]], i32 0, 1
+; CHECK-NEXT:    [[TMP3]] = insertvalue [2 x i32] [[TMP2]], i32 0, 1
+; CHECK-NEXT:    [[EXTRACT_VALUE:%.*]] = extractvalue [2 x i32] [[TMP3]], 0
+; CHECK-NEXT:    br label %[[EXIT]]
+; CHECK:       [[EXIT]]:
+; CHECK-NEXT:    [[SINK_VALUE:%.*]] = phi i32 [ 0, %[[ENTRY]] ], [ [[EXTRACT_VALUE]], %[[UNREACHABLE_BLOCK]] ]
+; CHECK-NEXT:    ret i32 0
+;
+entry:
+  br label %exit
+
+unreachable.block:                                ; No predecessors!
+  %0 = insertvalue [2 x i32] %3, i32 0, 1
+  %1 = insertvalue [2 x i32] %0, i32 0, 1
+  %2 = insertvalue [2 x i32] %1, i32 0, 1
+  %3 = insertvalue [2 x i32] %2, i32 0, 1
+  %extract.value = extractvalue [2 x i32] %3, 0
+  br label %exit
+
+exit:                             ; preds = %unreachable.block, %entry
+  %sink.value = phi i32 [ 0, %entry ], [ %extract.value, %unreachable.block ]
+  ret i32 0
+}
-- 
2.53.0.1213.gd9a14994de-goog

